flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/snapshot/StateSnapshotReconciler.java

```diff
@@ -0,0 +1,235 @@
+import org.apache.flink.configuration.CheckpointingOptions;
+import org.apache.flink.kubernetes.operator.api.CrdConstants;
+import org.apache.flink.kubernetes.operator.api.FlinkDeployment;
+import org.apache.flink.kubernetes.operator.api.FlinkStateSnapshot;
+import org.apache.flink.kubernetes.operator.api.spec.FlinkStateSnapshotSpec;
+import org.apache.flink.kubernetes.operator.api.status.FlinkStateSnapshotState;
+import org.apache.flink.kubernetes.operator.api.status.SnapshotTriggerType;
+import org.apache.flink.kubernetes.operator.controller.FlinkResourceContext;
+import org.apache.flink.kubernetes.operator.controller.FlinkStateSnapshotContext;
+import org.apache.flink.kubernetes.operator.reconciler.ReconciliationUtils;
+import org.apache.flink.kubernetes.operator.service.FlinkResourceContextFactory;
+import org.apache.flink.util.Preconditions;
+
+import io.javaoperatorsdk.operator.api.reconciler.DeleteControl;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.ObjectUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Optional;
+
+/** The reconciler for the {@link org.apache.flink.kubernetes.operator.api.FlinkStateSnapshot}. */
+@RequiredArgsConstructor
+public class StateSnapshotReconciler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StateSnapshotReconciler.class);
+
+    private final FlinkResourceContextFactory ctxFactory;
+
+    public void reconcile(FlinkStateSnapshotContext ctx) throws Exception {
+        var source = ctx.getResource().getSpec().getJobReference();
+        var resource = ctx.getResource();
+
+        var savepointState = resource.getStatus().getState();
+        if (!FlinkStateSnapshotState.TRIGGER_PENDING.equals(savepointState)) {
+            return;
+        }
+
+        if (resource.getSpec().isSavepoint()
+                && resource.getSpec().getSavepoint().getAlreadyExists()) {
+            LOG.info(
+                    "Snapshot {} is marked as completed in spec, skipping triggering savepoint.",
+                    resource.getMetadata().getName());
+            resource.getStatus().setState(FlinkStateSnapshotState.COMPLETED);
+            resource.getStatus().setPath(resource.getSpec().getSavepoint().getPath());
+            var time = DateTimeUtils.kubernetes(Instant.now());
+            resource.getStatus().setTriggerTimestamp(time);
+            resource.getStatus().setResultTimestamp(time);
+            return;
+        }
+
+        var secondaryResource =
+                ctx.getSecondaryResource()
+                        .orElseThrow(
+                                () ->
+                                        new RuntimeException(
+                                                String.format(
+                                                        "Secondary resource %s not found",
+                                                        source)));
```

Comment:
Should this just abandon the snapshot?

Reply:
This feels a bit weird, since after the observe phase, the code should not be able to run into this at all. I think it would help keep the code cleaner if we only abandon snapshots from the observer.